mantidproject/mantid
# Muon - Load Run Widget (Single Thread)

closes #23323 , requires #23371 to be merged before merging this.

This PR replaces #23334 by removing multi-threading functionality.

This PR creates a new "load run widget" for use in the new Muon GUI. See the issue for a description of the main features.

Essentially the widget allows for a user to load a single or multiple files via their run numbers, via a *Load Current Run* button or via explicitly entering the runs into the regex protected `QLineEdit`.

Additionally, there are two buttons "<" and ">" (or "decrement run" and "increment run"). The behaviour of these varies depending on whether the mode is set to single file or multiple file. In single file mode the increment/decrement changes the run number by one, loads that run number and overwrites the previous data. In multiple file mode the new run is loaded and appending to the existing set of loaded runs.

Some files are duplicated from PR #23371 and **do not** need review here (mainly utility functions)

- `muon_file_utils.py`
- `run_string_utils.py`
- `muon_load_data.py`

Testing
=======

The unit tests on the presenter are a good reference point for the functionality I have explicitly coded into the widget.

The widget can be tested without Mantid using the following test code (for example add a file `test.py` into `/scripts/`)

```python
from __future__ import (absolute_import, division, print_function)

from qtpy import QtGui, QtCore, QtWidgets
from mantid.simpleapi import *
from Muon.GUI.MuonAnalysis.loadrun.load_run_model import BrowseRunWidgetModel
from Muon.GUI.MuonAnalysis.loadrun.load_run_view import BrowseRunWidgetView
from Muon.GUI.MuonAnalysis.loadrun.load_run_presenter import BrowseRunWidgetPresenter

from Muon.GUI.Common.muon_load_data import MuonLoadData

if __name__ == "__main__":
    import sys

    app = QtWidgets.QApplication(sys.argv)
    ui = BrowseRunWidgetPresenter(BrowseFRunWidgetView(), BrowseRunWidgetModel(MuonLoadData()))
    ui.show()
    sys.exit(app.exec_())
```

Which will display the widget.

## Testing the functionality

Please make a general test of the widget in any way you see fit; causing any sort of crash or unexpected behaviour is a failure unless a sensible warning message is displayed.

A test on at least two operating systems would be great. I have developed on Windows, but all the functionality (including loading the current run) should work on Linux/Mac.

Specific behaviour to test:

- Try to cause a crash in any way possible.

- Try to load multiple runs by using the line edit. Check the console output to ensure the Load algorithm is being run. Check that the syntax makes sense (you can only enter digits or "," and "-" as delimiter and range separator respictvely)

- Try to load a run using *Load Current Run* (make sure you are on the ISIS network). Currently this is set up for the EMU instrument.

- Check that the increment/decrement buttons work and their behaviour is intuitive.

- Switch between single/multiple file mode and repeat the above tests. To switch to multiple file mode, just add the line `self.presenter.enable_multiple_files(True)` to the test code above.

## Code Review

- Is the code readable and commented well, are functions names understandable?
- Are the units tests appropriate, and do they cover the main functionality. Try to think of any functionality not covered by a test.
- Is the structure appropriate, e.g. using MVP correctly.

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
